pjmedia/src/pjmedia/port.c

```diff
@@ -124,17 +124,94 @@ PJ_DEF(pj_status_t) pjmedia_port_put_frame( pjmedia_port *port,
  */
 PJ_DEF(pj_status_t) pjmedia_port_destroy( pjmedia_port *port )
 {
-    pj_status_t status;
-
     PJ_ASSERT_RETURN(port, PJ_EINVAL);
 
+    if (port->grp_lock) {
+	pjmedia_port_dec_ref(port);
+	return PJ_SUCCESS;
+    }
+
+    if (port->on_destroy) {
+	return port->on_destroy(port);
+    }
+
+    return PJ_SUCCESS;
+}
+
+
+/* Group lock handler */
+static void port_on_destroy(void *arg)
+{
+    pjmedia_port *port = (pjmedia_port*)arg;
     if (port->on_destroy)
-	status = port->on_destroy(port);
-    else
-	status = PJ_SUCCESS;
+	port->on_destroy(port);
+}
+
+
+/**
+ * Create and init group lock.
+ */
+PJ_DEF(pj_status_t) pjmedia_port_init_grp_lock( pjmedia_port *port,
+						pj_pool_t *pool,
+						pj_grp_lock_t *glock )
+{
+    pj_grp_lock_t *grp_lock = glock;
+    pj_status_t status;
+
+    PJ_ASSERT_RETURN(port && pool, PJ_EINVAL);
+    PJ_ASSERT_RETURN(port->grp_lock == NULL, PJ_EEXISTS);
+
+    /* We need to be caution on ports that do not have the on_destroy()!
+     * It can be uninitialized yet or it really does not have one.
+     * If it doesn't have one, then we'd expect a possible premature destroy!
+     */
```

Comment:
Better print some logs so we can know during log analysis which port doesn't have on_destroy().